Cs/feature/algolia search reapply

cesarsanchez1 opened the pull request:
### _**Setup & Testing Guide**_

### **1) Local Environment Setup**

To test this branch locally, follow these steps:

**Check out this branch:**

- git fetch origin
- git checkout cs/feature/algolia-search-reapply

**Create a .env file in the project root (same directory as manage.py):**

- touch .env

**Add the following variables to your .env file — these are required for the Algolia integration:**

ALGOLIA_APP_ID=YourAlgoliaAppID
ALGOLIA_API_KEY=YourAlgoliaAdminAPIKey
ALGOLIA_SEARCH_KEY=YourAlgoliaSearchOnlyKey
ALGOLIA_INDEX_PREFIX=simpletix

**Make sure your virtual environment is active:**

- source venv/bin/activate


**Install dependencies:**

- pip install -r requirements.txt

**Run the server:**

- python3 manage.py runserver


### **2.) Setting Up an Algolia Account**

Go to https://www.algolia.com
 and create a free account.

Inside the Algolia dashboard, go to API Keys:

Copy your Application ID → use as ALGOLIA_APP_ID

Copy your Admin API Key → use as ALGOLIA_API_KEY

Copy your Search-Only API Key → use as ALGOLIA_SEARCH_KEY

Go to Indices → Create Index and create one named:

- simpletix_simpletix_events


This ensures alignment with Django’s backend sync.


### **3.) Running Locally**

Once .env and Algolia are configured:

- python3 manage.py migrate
- python3 manage.py runserver


Then navigate to:
http://127.0.0.1:8000/

### **4.) Testing the Integration**

**Log in to a Organizer Account.**

Go to Events → Create Event, and submit a test event.

Open your Algolia dashboard → Indices, and confirm the new event appears under simpletix_simpletix_events.

Return to your site and type the event title into the navbar search bar, the dropdown should update in real-time with your new event.

CRUD operations (Create, Read, Update, Delete) automatically sync with Algolia.
The search dropdown uses algoliasearch to query the same index dynamically.

**Result:** 

- Fully functioning CRUD + search integration
- Works locally and is AWS Secrets compatible
- Stable, consistent index naming

cesarsanchez1 commented:
That's it as far as deployment, down below I respond to some of the comments left by @thaddeuspearson 




### **Implementation Details**

The Algolia integration is now environment-aware and partially dynamic.

All credentials (ALGOLIA_APP_ID, ALGOLIA_API_KEY, ALGOLIA_SEARCH_KEY, ALGOLIA_INDEX_PREFIX) are dynamically loaded from .env or AWS Secrets Manager — no hardcoded values remain.

Index naming is partially static to ensure full synchronization between Django and the frontend.

Django dynamically builds the prefix from ALGOLIA_INDEX_PREFIX, while the _events suffix remains fixed for stability.

This results in consistent indexing (e.g., simpletix_simpletix_events) across environments without introducing search mismatches.

The configuration is AWS-compatible, allowing production and development to use the same Secrets Manager key structure.


### **Index Naming Rationale**

The index_name in algolia_index.py is intentionally set to 'simpletix_events'.

During testing, a fully dynamic prefix-based approach caused a mismatch where Django wrote to simpletix_simpletix_events while the frontend queried simpletix_events.

To resolve this, the index name was standardized to 'simpletix_events', ensuring Django CRUD operations and the frontend search dropdown reference the same index.

The prefix simpletix_ (from ALGOLIA_INDEX_PREFIX=simpletix) keeps backend and frontend aligned without breaking functionality.

This balanced approach preserves flexibility for future dynamic handling while guaranteeing consistent indexing behavior across all environments.


### **Regarding @thaddeuspearson’s suggestion** to standardize constant names (e.g., ALGOLIA_APP_ID vs. APPLICATION_ID):
This implementation preserves the current key naming convention to remain compatible with the algoliasearch_django library, which expects APPLICATION_ID and API_KEY.

Standardizing everything to the ALGOLIA_ prefix would break backend indexing without modifying the library’s internal configuration or duplicating keys.

Since all environment variables already align with AWS Secrets Manager (ALGOLIA_APP_ID, ALGOLIA_API_KEY, etc.), this approach maintains both library compatibility and AWS consistency.

thaddeuspearson commented on simpletix/config/context_processors.py:
```diff
@@ -0,0 +1,8 @@
+
+def algolia_settings(request):
+    return {
+        "ALGOLIA_APP_ID": settings.ALGOLIA.get("APPLICATION_ID", ""),
```
See my note I left on  settings.py.

These constants will likely need to be updated to:

```python
"ALGOLIA_APP_ID": settings.ALGOLIA.get("ALGOLIA_APPLICATION_ID", ""),
"ALGOLIA_SEARCH_KEY": settings.ALGOLIA.get("ALGOLIA_SEARCH_KEY", ""),
"ALGOLIA_INDEX": f"{settings.ALGOLIA.get('ALGOLIA_INDEX_PREFIX', 'simpletix')}_events",
```

to keep things consistent for debugging.

Alternatively, we _could_ store `"ALGOLIA_INDEX"` directly in secrets manager, so we don't need to concatenate it manually here. My preferance that we change it, since we only use `"ALGOLIA_INDEX_PREFIX"` to create `"ALGOLIA_INDEX"` manually, but I am open for discussion about this

thaddeuspearson reviewed (changes requested):
This is looking good, but there are a couple of concerns that need to be addressed before we can merge this in.

thaddeuspearson commented on simpletix/simpletix/.DS_Store:
same thing here. We should not commit DS_Stores to source

cesarsanchez1 commented:
**Changes Implemented**

**1. Removed all `__pycache__` directories**
These are now cleaned out and ignored through `.gitignore` for future commits.

**2. Merged latest `develop` updates**
Branch is now fully up to date and conflict-free.

**3. Simplified environment loading**
Removed the redundant `from dotenv import load_dotenv` and now just use a single `load_dotenv()` since `.env` is already at the project root.

**4. Cleaned up old comments**
Removed the outdated lines in `algolia_index.py` and `models.py` as suggested.

**5. Media directory added to `.gitignore`**
Removed the temporary banner test files and added `media/` to prevent re-commits.

**6. Added environment-aware Algolia configuration**
The `settings.py` logic now handles both local and deployed environments:

```python
if ENVIRONMENT in ["production", "development"]:
    ALGOLIA = get_secret(os.getenv("ALGOLIA_SECRETS_NAME"))
else:
    ALGOLIA = {
        "APPLICATION_ID": os.getenv("ALGOLIA_APP_ID", ""),
        "API_KEY": os.getenv("ALGOLIA_API_KEY", ""),
        "SEARCH_KEY": os.getenv("ALGOLIA_SEARCH_KEY", ""),
        "INDEX_PREFIX": os.getenv("ALGOLIA_INDEX_PREFIX", "simpletix"),
    }
```

This setup lets local testing pull from `.env` while AWS deployments automatically pull from Secrets Manager.

**7. Context processor alignment**
The context processor now exposes:

```python
"ALGOLIA_APP_ID": settings.ALGOLIA.get("APPLICATION_ID", ""),
"ALGOLIA_SEARCH_KEY": settings.ALGOLIA.get("SEARCH_KEY", ""),
"ALGOLIA_INDEX": f"{settings.ALGOLIA.get('INDEX_PREFIX', 'simpletix')}_events",
```

I kept the `APPLICATION_ID` / `API_KEY` naming internally since those are the constants expected by the `algoliasearch_django` library.
This ensures backend CRUD syncing works correctly while `.env` and AWS Secrets still use the prefixed names (`ALGOLIA_APP_ID`, etc.).

**8. Verified single `Ticket` model usage**
Confirmed there’s no duplication, only one valid `Ticket` model remains.



### 
* The Algolia integration is now stable, environment-aware, and fully compatible with AWS Secrets Manager.
* CRUD operations sync correctly to `simpletix_simpletix_events`.
* Frontend search dropdown updates in real time and queries the same index.
* Local and deployed setups require no code change, only environment variable configuration.

thaddeuspearson commented:
@cesarsanchez1 great work bringing this PR home! Once @xl2024 's [PR](https://github.com/gcivil-nyu-org/team1-mon-fall25/pull/96#pullrequestreview-3381059244) is merged in, you can pull in the changes from `develop` and we will get his merged in too!
